### PR TITLE
Gmarchetti/transfer xchain to pchain test

### PR DESCRIPTION
**Pull request by galenmarchetti**
MERGE AFTER #49 (branched off of that PR)

### CODE REVIEW

File: commons/ava_testsuite/five_node_staking_gecko_network_tests.go [hunk 1]
```diff
@@ -1,14 +1,84 @@
...
 	"github.com/sirupsen/logrus"
+	"strconv"
 	"time"
 )
 
+const (
+	USERNAME = "test"
+	PASSWORD = "test34test!23"
+	// defined in Gecko codebase for default genesis block
+	// PREFUNDED_ADDRESS = "6Y3kysjF9jnHnYkdS9yGAuoHyae2eNmeV"
```
Comment (mieubrisse) [inline commons/ava_testsuite/five_node_staking_gecko_network_tests.go:18]: Why is this commented out?

File: commons/ava_testsuite/five_node_staking_gecko_network_tests.go [hunk 2]
```diff
@@ -1,14 +1,84 @@
...
+)
+
+type FiveNodeStakingNetworkPChainImportTest struct{}
+func (test FiveNodeStakingNetworkPChainImportTest) Run(network interface{}, context testsuite.TestContext) {
+	castedNetwork := network.(fixed_gecko_network.FixedGeckoNetwork)
+	testAmount := 10000
```
Comment (mieubrisse) [inline commons/ava_testsuite/five_node_staking_gecko_network_tests.go:24]: can this be a const?

File: commons/ava_testsuite/five_node_staking_gecko_network_tests.go [hunk 3]
```diff
@@ -1,14 +1,84 @@
...
+	referenceNodeClient, err := castedNetwork.GetGeckoClient(4)
```
Comment (mieubrisse) [inline commons/ava_testsuite/five_node_staking_gecko_network_tests.go:25]: why specifically? should be a named const

File: gecko_client/keystore_user_types.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package gecko_client
+
+type CreateUser struct {
+	Success bool `json:"success"`
+}
+
+type CreateUserResponse struct {
```
Comment (mieubrisse) [inline gecko_client/keystore_user_types.go:7]: Noice job breaking this into its own file!

File: gecko_client/xchain_api.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	var response ImportKeyResponse
+	if err := json.Unmarshal(responseBodyBytes, &response); err != nil {
+		return "", stacktrace.Propagate(err, "Error unmarshalling JSON response")
+	}
+	return response.Result.Address, nil
+}
+
+func (api XChainApi) ExportAVA(to string, amount int, username string, password string) (string, error) {
```
Comment (mieubrisse) [inline gecko_client/xchain_api.go:34]: Amounts should be int64 since it's specified in nAva

File: gecko_client/xchain_api.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	}
+
+	var response GetBalanceResponse
+	if err := json.Unmarshal(responseBodyBytes, &response); err != nil {
+		return nil, stacktrace.Propagate(err, "Error unmarshalling JSON response")
+	}
+	return &response.Result, nil
+}
+
+func (api XChainApi) Send(amount int, assetId string, to string, username string, password string) (string, error) {
```
Comment (mieubrisse) [inline gecko_client/xchain_api.go:86]: Amount should be int64

File: gecko_client/xchain_types.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	Id             int                     `json:"id"`
+}
+
+type UtxoIdInfo struct {
+	TxID string `json:"txID"`
+	OutputIndex int `json:"outputIndex"`
+}
+
+type AccountWithUtxoInfo struct {
+	Balance string `json:"balance"`
```
Comment (mieubrisse) [inline gecko_client/xchain_types.go:25]: This should be an int64, not a string, right?

Comment (mieubrisse) [reply gecko_client/xchain_types.go:25]: Oh wtf, their API seems messed up

File: gecko_client/xchain_api.go [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+const (
+	xchainEndpoint = "ext/bc/X"
+)
+
+type XChainApi struct {
+	rpcRequester jsonRpcRequester
+}
+
+func (api XChainApi) ImportKey(username string, password string, privateKey string) (string, error) {
```
Comment (mieubrisse) [inline gecko_client/xchain_api.go:16]: We're adding a bunch of endpoints but no tests to verify that it works as expected. I'm happy to do this; just give me the greenlight and I'll create a branch off this branch, generate a bunch of tests for these guys, and submit a PR against yours

File: commons/ava_testsuite/rpc_manager.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+const (
+	GENESIS_USERNAME = "genesis"
+	GENESIS_PASSWORD = "genesis34!23"
+	ACCEPTED_STATUS = "Accepted"
+)
+
+type RpcManager struct {
+	client *gecko_client.GeckoClient
+	testNet *ava_default_testnet.TestNet
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:19]: I think we have two ways forward here:
1) don't pass in the TestNet as a parameter, since there's only one TestNet and it's a constant (`DefaultTestNet`) so just use that at all points where you'd normally have used `rpcManager.testNet` (my preferred way)
2) Emphasize that the `TestNet` type is generic by moving it to its own file (`ava_test_net` maybe) and that `ava_default_testnet` is but one implementation. I prefer this less because, to my understanding, we're always going to roll with DefaultTestNet

File: commons/ava_testsuite/rpc_manager.go [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+	username string
+	password string
+	payerNonce int
+}
+
+func NewRpcUser(username string, password string) *RpcUser {
+	return &RpcUser{username: username, password: password, payerNonce: 0}
+}
+
+func (rpcUser RpcUser) incrementNonce() int {
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:45]: wat dis do?

File: commons/ava_testsuite/rpc_manager.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+/*
+	Creates a new account on the XChain under the username and password.
+	Transfers funds from the genesis account to the new XChain account using the Genesis private key.
+	Returns the new, funded XChain account address.
+ */
+func (rpcManager RpcManager) CreateAndSeedXChainAccountFromGenesis(
+	amount int) (string, error) {
+	time.Sleep(time.Second * 30)
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:57]: Whoa why is this here?

File: commons/ava_testsuite/rpc_manager.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/kurtosis-tech/ava-e2e-tests/gecko_client"
+	"github.com/palantir/stacktrace"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+const (
+	GENESIS_USERNAME = "genesis"
+	GENESIS_PASSWORD = "genesis34!23"
+	ACCEPTED_STATUS = "Accepted"
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:14]: Let's add a bit more info about what exactly is accepted (a transaction? a user? etc.). General pattern I try to follow: rather than describing what the constant's value is, you describe what the constant's semantic meaning is.

File: commons/ava_testsuite/rpc_manager.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+		return "", stacktrace.Propagate(err, "Could not get node id")
+	}
+	genesisAccountAddress, err := client.XChainApi().ImportKey(
+		GENESIS_USERNAME,
+		GENESIS_PASSWORD,
+		rpcManager.testNet.FundedAddresses.PrivateKey)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed to take control of genesis account.")
+	}
+	logrus.Debugf("Adding Node %s as a validator.", nodeId)
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:80]: Fun fact: if you use `%v` you don't have to think about whether it's a string, list, int, or whatever

File: commons/ava_testsuite/rpc_manager.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	logrus.Debugf("Genesis Address: %s.", genesisAccountAddress)
+	testAccountAddress, err := client.XChainApi().CreateAddress(username, password)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed to create address on XChain.")
+	}
+	logrus.Debugf("Test account address: %s", testAccountAddress)
+	txnId, err := client.XChainApi().Send(amount, "AVA", testAccountAddress, GENESIS_USERNAME, GENESIS_PASSWORD)
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:87]: No magic strings please - let's make this a const (something like `AVA_ASSET_ID` or something)

File: commons/ava_testsuite/rpc_manager.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	_, err = client.KeystoreApi().CreateUser(GENESIS_USERNAME, GENESIS_PASSWORD)
+	if err != nil {
+		stacktrace.Propagate(err, "Could not create genesis user.")
+	}
+	nodeId, err := client.AdminApi().GetNodeId()
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Could not get node id")
+	}
+	genesisAccountAddress, err := client.XChainApi().ImportKey(
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:73]: Bookmark to chat this over: would love to understand how the accounts & keys work; don't feel like I have a strong enough knowledge right now to comment

File: commons/ava_testsuite/rpc_manager.go [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+	"time"
+)
+
+const (
+	GENESIS_USERNAME = "genesis"
+	GENESIS_PASSWORD = "genesis34!23"
+	ACCEPTED_STATUS = "Accepted"
+)
+
+type RpcManager struct {
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:17]: Can we be more specific about what this guy does? It _seems_ like what he's really doing is bootstrapping a PChain account with Ava

File: commons/ava_testsuite/five_node_staking_gecko_network_tests.go [hunk 17]
```diff
@@ -118,10 +189,56 @@ func (test FiveNodeStakingNetworkGetValidatorsTest) Run(network interface{}, con
...
-	return fixed_gecko_network.NewFixedGeckoNetworkLoader(5, 1, true)
+	return fixed_gecko_network.NewFixedGeckoNetworkLoader(5, 5, true)
 }
 
 func (test FiveNodeStakingNetworkGetValidatorsTest) GetTimeout() time.Duration {
 	return 30 * time.Second
 }
 
+
+func createAndSeedXChainAccountFromGenesis(client *gecko_client.GeckoClient, username string, password string, amount int) (string, error) {
```
Comment (mieubrisse) [inline commons/ava_testsuite/five_node_staking_gecko_network_tests.go:200]: This looks like a copy-paste of the same method inside RpcManager. Is that intentional??

File: commons/ava_testsuite/rpc_manager.go [hunk 18]
```diff
@@ -0,0 +1,147 @@
+package ava_testsuite
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:1]: Thoughts about this file as a whole: AFAICT it looks like:
1. These methods are meant to be run in conjunction, one after each other
2. They're meant to be run exactly once
3. They're only used in one place

Since this is true, I don't think we want to have this logic as a separate struct - instead, it seems like we'd want to have a `initializePChainAccount` function in the 5-node staking network tests, and that function should do the equivalent of both of these.

File: commons/ava_testsuite/rpc_manager.go [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	txnId, err = client.PChainApi().ImportAVA(username, password, pchainAddress, rpcManager.rpcUser.incrementNonce())
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed import AVA to pchainAddress %s", pchainAddress)
+	}
+	_, err = client.PChainApi().IssueTx(txnId)
+	if err != nil {
+		return "", stacktrace.Propagate(err, "Failed to issue importAVA transaction.")
+	}
+	time.Sleep(time.Second * 5)
```
Comment (mieubrisse) [inline commons/ava_testsuite/rpc_manager.go:128]: Doesn't seem like we should be sleeping here - instead using a more precise method

File: commons/ava_testsuite/five_node_staking_gecko_network_tests.go (hunk 1 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/five_node_staking_gecko_network_tests.go:18]: no longer necessary, removing it

File: commons/ava_testsuite/rpc_manager.go (hunk 10 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go:46]: I'm going to refactor the nonce increments so they're tied to accounts

File: commons/ava_testsuite/rpc_manager.go (hunk 16 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go:18]: this.basically provides a much higher-level API than the underlying RPC calls that provides simpler language for normal testing tasks on a testnet

File: commons/ava_testsuite/rpc_manager.go (hunk 10 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go]: done

File: commons/ava_testsuite/rpc_manager.go (hunk 9 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go:21]: I'm going to move forward with #2 and do that refactor

File: gecko_client/xchain_api.go (hunk 8 above)
Comment (mieubrisse) [reply gecko_client/xchain_api.go:16]: Dope added https://github.com/kurtosis-tech/ava-e2e-tests/issues/52 to track

File: commons/ava_testsuite/rpc_manager.go (hunk 9 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go]: did the first option

File: commons/ava_testsuite/rpc_manager.go (hunk 16 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go]: refactored to high level gecko client

File: commons/ava_testsuite/rpc_manager.go (hunk 18 above)
Comment (galenmarchetti) [reply commons/ava_testsuite/rpc_manager.go]: refactored to high level gecko client
